flow/directives.py

```diff
@@ -0,0 +1,315 @@
+    """Return the 'actual number of processes/threads to use.
+
+    We check the default value because when aggregation occurs we add the number
+    of MPI ranks and OMP_NUM_THREADS. If we always took the greater of the given
+    NP and ranks * threads then after aggregating we will inflate the number of
+    processors needed as (r1 * t1) + (r2 * t2) <= (r1 * r2 * t1 * t2) for
+    numbers greater than one.
+    """
+    if callable(value) or value != _NP_DEFAULT:
+        return value
+    nranks = directives.get('nranks', 1)
+    omp_num_threads = directives.get('omp_num_threads', 1)
+    if callable(nranks) or callable(omp_num_threads):
+        return value
+    else:
+        return max(value, max(1, nranks) * max(1, omp_num_threads))
+
+
+natural_number = OnlyType(int, postprocess=raise_below(1))
+# Common directives and their instantiation as _DirectivesItem
+NP = _DirectivesItem('np', natural_number, _NP_DEFAULT, finalize=finalize_np)
+NP.__doc__ = """
+The number of tasks expected to run for a given operation
+
+Expects a natural number (i.e. an integer >= 1). When querying
+this directive, introspection into the Directives is used to determine if the
+product of "nranks" or "omp_num_threads" is greater than the current value. The
+maximum of these two values is used.
+"""
+
+nonnegative_int = OnlyType(int, postprocess=raise_below(0))
+NGPU = _DirectivesItem('ngpu', nonnegative_int, 0)
+NGPU.__doc__ = """
+The number of GPU's to request for this operation.
+
+Expects a nonnegative integer.
+"""
+NRANKS = _DirectivesItem('nranks', nonnegative_int, 0)
+NRANKS.__doc__ = """
+The number of MPI ranks to use for this operation.
+
+Expects a nonnegative integer.
+"""
+
+OMP_NUM_THREADS = _DirectivesItem('omp_num_threads', nonnegative_int, 0)
+OMP_NUM_THREADS.__doc__ = """
+The number of OpenMP threads to use for this operation.
+
+Expects a nonnegative integer.
+"""
+
+
+def no_aggregation(v, o):
+    return v
+
+
+EXECUTABLE = _DirectivesItem('executable', OnlyType(str), sys.executable,
+                             no_aggregation, no_aggregation)
+EXECUTABLE.__doc__ = """
+The location in the file system to the executable to be used for this operation.
+
+Expects a string pointing to a valid executable file in the
+current file system.
+
+By default this should point to a Python executable (interpreter); however, if
+the :py:class:`FlowProject` path is set to an empty string, the executable can
+be the file path to an executable Python script.
+"""
+
+
+nonnegative_real = OnlyType(float, postprocess=raise_below(0))
+WALLTIME = _DirectivesItem('walltime', nonnegative_real, 12.,
+                           lambda x, y: x + y, max)
+WALLTIME.__doc__ = """
+The number of hours to request for executing this job.
+
+This directive expects a float representing the number hours including
+fraction's of hours.
+"""
+
+MEMORY = _DirectivesItem('memory', natural_number, 4)
+MEMORY.__doc__ = """
+The number of GB to request for an operation.
```

Comment:
```suggestion
The number of gigabytes of memory to request for this operation.
```

Reply:
We need to be consistent about whether we say the resource is "requested" or "used" for each of these directives. I don't have a strong preference, just needs to match.